README.md
# Hidenza
A Haskell partial implementation of [Tyler Hobbs's Fidenza Algorithm](https://tylerxhobbs.com/fidenza)

## Motivation
Not being into NFTs at all, I've only very recently found out about the Fidenza algorithm, when it was mentioned as a prompt in [Genuary](https://genuary.art/) and I thought it looked wicked, so I read both the [flow field](https://tylerxhobbs.com/essays/2020/flow-fields) and [Fidenza](https://tylerxhobbs.com/fidenza) essays on Tyler's website in one breath. For the next couple of days I would find myself thinking about the ideas he mentioned and appreciating more and more how simple and yet clever they were.

That, paired with me only recently getting into this amazing programming language called [Haskell](https://www.haskell.org/) culminated into what you'll find in this repo.

## Quick run
In the `Usage` section below you can find information about all the arguments, but for the impatient ones, this should be enough to get the default arguments image, saved as `Fidenza.png`

`cabal run Fidenza -O2`

<p align="middle">
  <img alt="Resulting image of default parameters" src="github_imgs/default.png" width="50%"/>
</p>

I've also included a python script - `makeRandomImages.py`, which generates random seeds and creates images from them, so you can leave that running for a while to produce a bunch of images. It was super exciting for me to come back half an hour later and look at everything that's been produced and every time there would be at least a few surprising ones.

Here are a few of my favourite ones

<p float="left" align="middle">
  <img alt="" src="github_imgs/favourite_01.png" width="40%" vspace=10 hspace=20 />
  <img alt="" src="github_imgs/favourite_02.png" width="40%" vspace=10 /> 
  <img alt="" src="github_imgs/favourite_03.png" width="40%" vspace=10 hspace=20 />
  <img alt="" src="github_imgs/favourite_04.png" width="40%" vspace=10 />
</p>

## Disclaimer
There's a couple of things I would like to get out of the way as soon as possible.

### Haskell beginner
Not going into too much depth here, I've only very recently gotten into Haskell and the functional programming paradigm, so I still find it quite hard to move away from the imperative mindset, which leads to some quite sloppy code that you can find here.

### Partial implementation
I say **partial** implementation of the algorithm, as I am missing 2 of the parameters defined in the original Fidenza - spirals and density - and 4 of the colour schemes - dark lifestyle, rose, party time and white mono.

The **spirals** argument is missing, as I couldn't figure out the exact spiral equation. The closest I've gotten (which is not super close as the center still looks quite different) is  

```
a / (theta ^ 0.7), where a = 100
                         0 < theta <= 14 * Pi
```

which is then rotated a bunch of times to fill in the whole image.  
  
You'll find that I do have some images including spirals and those have been produced using the above method. What I did was, take those spirals and convert them to a vector field, which I then passed in to the algorithm using the `FromFile` constructor for the `vectorFieldGenerator` argument, which you can find below. I've stored those vectors in the `spiral.vectors` file if anyone would like to give them a go.

<p align="middle">
  <img alt="Example of a spiral image" src="github_imgs/spiral.png" width="50%"/>
</p>

The other one is **density**. Judging by the flow fields essay, I presume that circle packing is used when generating the Fidenza curves, but my approach is significantly different and way lazier. I literally try to generate a new completely random curve up to `numGenerationAttempts` times, until one manages to survive enough iterations to get longer than the `minLength` argument. If that sounds like throwing shit at the wall until something sticks, it's because that's the exact mathetmatical intuition behind it.

The result of that generation method is that the density of the ribbons is a bit of a magical combination between the `widths`,`fertilities`,`width`,`height`,`collisionMargin`,`maxCurves`,`minLength` and `numGenerationAttempts` arguments.

For the most part, I've been able to recreate the various scales present in the original Fidenzas without too much tweaking and those are stored as presets in the `scale` simple argument, which you can also read about below.

| microUniform | small
| --- | --- 
| <img alt="microUniform scale" src="github_imgs/microUniformScale.png" width=180px /> | <img alt="small scale" src="github_imgs/smallScale.png" width=180px /> 

| medium | uniform
| --- | ---
| <img alt="medium scale" src="github_imgs/mediumScale.png" width=180px /> | <img alt="uniform scale" src="github_imgs/uniformScale.png" width=180px />

| large | jumbo
| --- | --- | ---
| <img alt="large scale" src="github_imgs/largeScale.png" width=180px /> | <img alt="jumbo scale" src="github_imgs/jumboScale.png" width=180px />

| jumboXL
| ---
| <img alt="jumboXL scale" src="github_imgs/jumboXLScale.png" width=180px />

## Usage
Below is the help string for the executable which contains all the info you need for controlling your Fidenzas.

The section about _Simple Args_ contains the parameters as defined in the original Fidenza and provide the easiest way of controlling the algorithm.

```
A Haskell partial implementation of Tyler Hobbs's Fidenza algorithm.

Partial, as there currently is no support for creating spiral Fidenzas and for intuitive,
deterministic control over the density of Fidenzas.

The control over the density is achieved by tweaking the maximum number of curves and
the curve widths distribution, as if the image is already mostly filled up by shapes
and {numGenerationAttempts} attemps have failed to generate a new random ribbon, then
the algorithm ends.

A lot of arguments are of type distribution, which accepts 3 different kind of parameters:
   - a uniform distribution, specified by a list like so - [1,2,3,4,5,6], where
     each element has an equal probability of being chosen
   - a weighted distribution, specified by making each element a tuple like so:
     [(1,0.5),(2,0.1),(3,0.1),(4,0.1),(5,0.1),(6,0.1)], where the first member of
     each element is the value and the second is the probability of that value being chosen
   - a uniform distribution, specified by a range like so - (1,6,2), where
     the first number is the starting number, the second number is the end number and
     the third number is to step to keep taking from the starting number until you get to
     the end number. So (1,6,2) becomes [1,3,5].

Usage: (./Fidenza|cabal run Fidenza -- ) [ARGS] 

Ex:
  ./Fidenza --seed 1
  cabal run Fidenza -- --seed 1

Args:
   --avgBlockSize - whether to make all segments on a single curve the same.
      The range is (0,1). Default 0.0

   --chunkSizes - the distribution of chunk/segment sizes. Default (2,12,1)

   --chunksOverlap - when splitting into segments, we can choose to make them overlap a bit
     or choose to have a negative number which makes gaps between the segments. Default 1

   --collisionMargin - minimal distance between shapes. Default 2

   --colourScheme - the colour scheme as an id. See the list of colours at the
      Simple colour arg below. 9,10 and 11 are my own colour schemes. Default 0

   --customBgColour - the background colour in the format (r,g,b,a) with 0-255 values.
      Overrides the background of any defined colour schemes..

   --customColours - a custom colour palette in the format (r,g,b,a,probability)
      with 0-255 values for r,g,b and a and any type of relative probability.
      Overrides any colour scheme parameters. Default []

   --drawSoftly - whether to draw each shape by drawing many tiny strokes,
      rather than filling the whole shape, resulting in a soft,
      brush like looking shapes. Default 0

   --fertilities - the distribution of fertilities (how long a curve can get). Default (50,400,25)

   --height - image height. Default 1200

   --help - shows this help message.

   --maxCurves - a maximum number of curves before execution stops. Default 200

   --maxSteps - a way of limiting the execution, where one step is considered 
      the extension of any curve by one segment of length {stepLength}. Default 100000

   --minLength - the minimum number of {stepLength} long segments to accept 
      as a valid generated curve. Default 15

   --numGenerationAttempts - how many attempts to make when generating a new 
      curve before giving up. Default 1000

   --outlineColour - the colour of any outlines. Default (10,10,10,255)

   --outlineSize - the size of any outlines. Default 1

   --padding - minimal distance from a shape to the borders of the image. Default 20

   --randomBoil - randomly pushes shape points with this amount. Default 0

   --randomBoilSeed - the seed for the random boil. Uses a separate seed,
      so we can make small, silly boiling animations. Default 0

   --rotationOffset - rotates the whole vector field by this parameter. Default 0

   --seed - seed for the random generator. Default 0

   --skewAngles - the distribution of skewing angles applied to the segments
      when drawing. One angle per curve. Default (-0.1,0.1,0.01)

   --softMaxStrokes - maximum number of tiny strokes per curve. Default 220

   --softNumStrokesWidthRatio - multiplies the width of a curve by this to
      define the number of strokes per shape. Default 1.8

   --softPerlinFreq - the frequency/turbulence of the soft strokes. Default 0.016

   --softPerlinOfs - the offset of the soft strokes vector field,
      to allow (x,y) values outside of the img dimensions
      to have valid results. Default -300

   --softRandomOfsAlongPerp - how much to randomly push the tiny strokes
      along the perpendicular (normal) of the curve. Default 2

   --softRandomOfsAlongV - how much to randomly push the tiny strokes along
      along the direction of the curve. Default 6

   --softSeed - seed for the softness, so it can be animated. Default 0

   --softStepLength - how many stes of {stepLen} to group into one step
      for the tiny strokes. Default 11

   --squareBlocks - whether to make the segments along the ribbon square.
      The range is (0,1). Default 0.0

   --stepLength - how long of a step to make at every iteration.
      Inveresely proportional to resolution. Default 1

   --stopChunkingAt - controls the distance from start and end to stop 
      chunking the ribbon into segments at. The actual number is randomly 
      sampled between the left and right boundaries.
      Set this to something massive if you don't want segments. Default (5,15)

   --strokeOrFill - whether to outline each segment (0), fill the each segment (1),
      outline and fill each segment (2) or fill each segment and
      outline the whole ribbon. Default 1

   --vectorFieldGenerator - method for generating the underlying vector field
      Valid options are:
         - PerlinNoise {frequency} {offset}, where the offset
              is there to allow us to pass negative coordinates
              and still get a result
         - FromFile {filePath}, where the file has the follwoing forma:
              x11 y11 x12 y12 x13 y13 .. x1n y1n
              x21 y21 x22 y22 x23 y23 .. x2n y2n
              ..
              xm1 ym1 xm2 ym2 xm3 ym3 .. xmn ymn

              the m and n numbers are not necessary to be the same as the
              width and height of the image as they will be lerped
         - Horizontal - every point in the vector field returns (1,0)
         - Straight {angle} - every point in the vector field retuns (1,0)
              rotated by {angle}. Default PerlinNoise 0.001 -400

   --vectorFieldStepPiDivisor - if greater than 0, all angles will be rounded to 
      multiples of PI divided by this parameter. Default 0

   --width - image width. Default 1000

   --widths - the distribution of curve widths. Default (2,60,4)


Simple Args:
  If the simple flags are used, then the arguments accepted are significantly
  simplified and act as switches on defined presets of the arguments above.

  These presets are designed to resemble Tyler Hobbs's parameters as much as possible.

  It's important to note that the presets are evaluated first, and then any of the main
  arguments above, which means preset values can be overwritten by the main arguments.

  For example, the following command will NOT use the 'small' scale as specified by the
  preset, but instead use the widths provided afterwards.

  ./Fidenza --scale small --widths  [80,90]

   --collisionCheck - whether or not to allow collisions
      Valid options are noOverlap, relaxed and anythingGoes.

   --colours - colour scheme
      Valid options are luxe,rad,golfSocks,baked,politique,
      cool,am,whiteOnCream,blackAndWhite and derived+(any of the above).
        e.g. derivedAm or derivedBlackAndWhite.

   --haveMargin - whether we are constrained to the frame or allowed to go outside.

   --outlined - whether to add an outline to each segment.

   --scale - defines the size distribution of the ribbons
      Valid options are small,medium,large,jumbo,jumboXL,microUniform and uniform.

   --shapeAngles - whether the turns in the ribbons are smooth and sharp
      Valid options are curved and sharp..

   --softShapes - whether to draw the shapes with hundreds of small strokes, 
      resulting in a soft, brush like looking stroke. Valid options are yes and no.

   --superBlocks - whether to make the segments along the ribbon square
      Valid options are yes and no.

   --turbulence - how drastic the shape turns are. A more technical explanation 
      is that this parameter controls the frequency of the underlying 
      perlin noise vector field. Valid options are none,low,mid and high.
```

## License
Of course, full credit for the algorithm goes to [Tyler Hobbs](https://tylerxhobbs.com/).

This repo, though, is licensed under GPLv3.


CHANGELOG.md
# Revision history for Fidenza

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


src/PerlinNoise.hs
{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}

module PerlinNoise (noise2d) where

import System.Random

type Vec2 = (Float,Float)

toVec2 :: (Int,Int) -> Vec2
toVec2 (x,y) = (fromIntegral x, fromIntegral y)

subVec2 :: Vec2 -> Vec2 -> Vec2
subVec2 (x1,y1) (x2,y2) = (x1-x2,y1-y2)

dotVec2 :: Vec2 -> Vec2 -> Float
dotVec2 (x1,y1) (x2,y2) = x1*x2 + y1*y2

getGrad :: Int -> StdGen -> (Int,Int) -> Vec2
getGrad w gen (x,y) = grads !! (y * w + x)
  where gradOptions = [(gx,gy) | gx <- [-1,1], gy <- [-1,1]]
        grads = map
                  (gradOptions !!)
                  $ randomRs (0, 3) gen

fade :: Float -> Float
fade t = t*t*t*(t*(t*6-15)+10)

lerp :: Float -> Float -> Float -> Float
lerp a b w = a + (b-a) * w

noise2d :: (Float,Float) -> Int -> StdGen -> Float
noise2d (x,y) w gen = raw * 0.5 + 0.5
  where (binX,binY) = (floor x, floor y)
        (xf,yf) = (x - fromIntegral binX, y - fromIntegral binY)
        verts = [(nx,ny) | nx <- [(binX),(binX+1)],
                           ny <- [(binY),(binY+1)]]
        grads = map (getGrad w gen) verts
        dists = map (subVec2 (x,y) . toVec2) verts
        [dpTL,dpBL,dpTR,dpBR] = zipWith dotVec2 grads dists
        (u,v) = (fade xf, fade yf)
        raw = lerp (lerp dpTL dpTR u)
                   (lerp dpBL dpBR u) v


src/Distribution.hs
module Distribution ( Distribution(..)
                    , toWeightedDistribution
                    , mapDistribution
                    , readDistribution
                    , sampleDistribution
                    , nullDistribution
                    ) where

import Data.List (sortBy)

data Distribution a = Distribution [(a,Float)] 

instance (Show a) => Show (Distribution a) where
    show (Distribution as) | length as > 5 =
      "Distribution [" ++ show (head as) ++ ".." ++ show (last as) ++ "]"
                           | otherwise     = "Distribution " ++ show as

mapDistribution :: ((a,Float) -> b) -> Distribution a -> [b]
mapDistribution f (Distribution as) = map f as

readDistribution :: (Enum a, Read a, Num a) => String -> Distribution a
readDistribution str@('[':'(':_) = toWeightedDistribution $ read str
readDistribution str@('[':_)     = toUniformDistribution $ read str
readDistribution str@('(':_)     = toUniformDistribution $ asList
  where (bottom,top,step) = read str
        asList = [bottom,(bottom+step)..top]
readDistribution _ = error "Can't parse distribution"

toWeightedDistribution :: [(a,Float)] -> Distribution a
toWeightedDistribution = Distribution . sort' . normalize'
  where sort' = sortBy (\a b -> compare (snd a) (snd b))
        normalize' xs = map (\(v,prob) -> (v,prob/sum')) xs
            where sum' = sum $ map snd xs

toUniformDistribution :: [a] -> Distribution a
toUniformDistribution as = Distribution $ zip as $ repeat prob
  where prob = 1.0 / (fromIntegral $ length as)

sampleDistribution :: Distribution a -> Float -> a
sampleDistribution (Distribution distAs) x =
  go x distAs
  where go _ [] = error "Can't sample empty distribution"
        go v ((a,prob):_)  | v <= prob  = a
        go v ((_,prob):as) | otherwise = go (v-prob) as

nullDistribution :: Distribution a -> Bool
nullDistribution (Distribution []) = True
nullDistribution _ = False


src/ColourSchemes.hs
module ColourSchemes ( colourSchemes
                     , readColour
                     , readColourProbs
                     ) where

import Codec.Picture( PixelRGBA8( .. ), Pixel8)
import Data.List.Split (splitOn)

import Distribution (Distribution(..), toWeightedDistribution)

type ColourScheme = (PixelRGBA8, Distribution PixelRGBA8)

luxe,rad,golfSocks,baked,politique :: ColourScheme
cool,am,whiteOnCream :: ColourScheme
colourSchemes :: [ColourScheme]

luxe = (PixelRGBA8 235 228 216 255,
        toWeightedDistribution [(PixelRGBA8  84  62  46 255, 0.08)
                               ,(PixelRGBA8 184 217 206 255, 0.06)
                               ,(PixelRGBA8 224 215 197 255, 0.05)
                               ,(PixelRGBA8  59  43  32 255, 0.04)
                               ,(PixelRGBA8 252 188  25 255, 0.03)
                               ,(PixelRGBA8 219  79  84 255, 0.03)
                               ,(PixelRGBA8 252 210 101 255, 0.02)
                               ,(PixelRGBA8  31  51  89 255, 0.02)
                               ,(PixelRGBA8  41 166 145 255, 0.02)
                               ,(PixelRGBA8  49  95 140 255, 0.02)
                               ,(PixelRGBA8 209  42  47 255, 0.02)
                               ,(PixelRGBA8 247 177 161 255, 0.01)
                               ,(PixelRGBA8 124 169 191 255, 0.01)
                               ,(PixelRGBA8 236 229 217 255, 0.01)
                               ,(PixelRGBA8 230 125  50 255, 0.01)])

rad = (PixelRGBA8 235 228 216 255,
       toWeightedDistribution [(PixelRGBA8 250 248 245 255, 0.33)
                              ,(PixelRGBA8 252 210 101 255, 0.06)
                              ,(PixelRGBA8 219  79  84 255, 0.04)
                              ,(PixelRGBA8 124 169 191 255, 0.03)
                              ,(PixelRGBA8 84  62   46 255, 0.02)
                              ,(PixelRGBA8 230 125  50 255, 0.02)
                              ,(PixelRGBA8 224 215 197 255, 0.02)
                              ,(PixelRGBA8 252 188  25 255, 0.01)])

golfSocks = (PixelRGBA8 102 128 106 255,
             toWeightedDistribution [(PixelRGBA8 231 225 213 255, 0.35)
                                    ,(PixelRGBA8 41 79 48 255, 0.16)
                                    ,(PixelRGBA8 36 77 115 255, 0.04)
                                    ,(PixelRGBA8 234 144 143 255, 0.03)
                                    ,(PixelRGBA8 34 75 115 255, 0.02)
                                    ,(PixelRGBA8 232 144 142 255, 0.01)
                                    ,(PixelRGBA8 22 50 77 255, 0.06)
                                    ,(PixelRGBA8 237 196 191 255, 0.01)])

baked = (PixelRGBA8  63 140  70 255,
         toWeightedDistribution [(PixelRGBA8  63 140  70 255, 0.59)
                                 ,(PixelRGBA8 247 212 203 255, 0.16)
                                 ,(PixelRGBA8 250 248 245 255, 0.10)
                                 ,(PixelRGBA8 202 230 204 255, 0.05)
                                 ,(PixelRGBA8 247 177 161 255, 0.04)
                                 ,(PixelRGBA8  99  77  58 255, 0.00)])

politique = (PixelRGBA8 235 228 216 255,
             toWeightedDistribution [(PixelRGBA8 252 210 101 255, 0.16)
                                    ,(PixelRGBA8 250 248 245 255, 0.20)
                                    ,(PixelRGBA8 247 177 161 255, 0.09)
                                    ,(PixelRGBA8  49  95 140 255, 0.03)
                                    ,(PixelRGBA8 219  79  84 255, 0.03)])
cool = (PixelRGBA8 235 228 216 255,
        toWeightedDistribution [(PixelRGBA8 128 171 192 255, 0.20)
                               ,(PixelRGBA8  58 101 143 255, 0.20)
                               ,(PixelRGBA8  34  54  91 255, 0.20)
                               ,(PixelRGBA8  21  30  54 255, 0.20)
                               ,(PixelRGBA8 224 215 197 255, 0.04)
                               ,(PixelRGBA8 203 193 224 255, 0.05)
                               ,(PixelRGBA8 185 217 206 255, 0.05)
                               ,(PixelRGBA8  44 167 146 255, 0.04)
                               ,(PixelRGBA8 208 230 221 255, 0.05)
                               ,(PixelRGBA8  61  45  34 255, 0.02)
                               ,(PixelRGBA8 251 210 103 255, 0.01)])
am = (PixelRGBA8  61  54  77 255,
      toWeightedDistribution [(PixelRGBA8  44  41  51 255, 0.50)
                              ,(PixelRGBA8 127 114 127 255, 0.15)
                              ,(PixelRGBA8 204 185 184 255, 0.15)
                              ,(PixelRGBA8 142 178 148 255, 0.04)
                              ,(PixelRGBA8 231 211 172 255, 0.04)
                              ,(PixelRGBA8 224 141 135 255, 0.04)])

whiteOnCream = (PixelRGBA8 235 228 216 255,
                toWeightedDistribution [(PixelRGBA8 250 248 245 255, 1.0)])

colourSchemes = [luxe, rad, golfSocks, baked, politique, cool, am, whiteOnCream]

-- Parsers
readColour :: String -> PixelRGBA8
readColour str = tupleToColour $ read str

readColourProbs :: String -> Distribution PixelRGBA8
readColourProbs [] = Distribution []
readColourProbs "[]" = Distribution []
readColourProbs ('[':str) =
    toWeightedDistribution $ map (readOne . (++")")) $ init $ splitOn ")" $ init str
    where readOne (',':str') = tupleToColourProb $ read str'
          readOne str' = tupleToColourProb $ read str'
readColourProbs str = error $ "Failed interpreting colour list " ++ str

tupleToColourProb :: (Pixel8,Pixel8,Pixel8,Pixel8,Float) -> (PixelRGBA8, Float)
tupleToColourProb (r,g,b,a,prob) = (PixelRGBA8 r g b a, prob)

tupleToColour :: (Pixel8,Pixel8,Pixel8,Pixel8) -> PixelRGBA8
tupleToColour (r,g,b,a) = PixelRGBA8 r g b a


src/VectorFieldGenerator.hs
module VectorFieldGenerator ( VectorFieldGenerator(..)
                            , VectorFieldFunc
                            , readVectorFieldFromFile
                            ) where

import Data.List (isPrefixOf)

data VectorFieldGenerator = PerlinNoise Float Int
                          | FromFile String
                            deriving stock (Show)

instance Read VectorFieldGenerator where
    readsPrec _ str
      | isPrefixOf "PerlinNoise " str =
          let params = words $ drop (length "PerlinNoise ") str
           in [(PerlinNoise (read $ params !! 0) (read $ params !! 1), "")]
      | isPrefixOf "FromFile " str =
          [(FromFile $ drop (length "FromFile ") $ str, "")]
    readsPrec _ str = error $ "VectorFieldGenerator.read: Can't read " ++ str

type Vector = (Float,Float)
type VectorField = [[Vector]]
type VectorFieldFunc = (Vector -> Vector)

readVectorFieldFromFile :: (Int,Int) -> FilePath -> IO (Vector -> Vector)
readVectorFieldFromFile (width,height) fp = do
    contents <- readFile fp
    let toTuples [] = []
        toTuples (_:[]) = error $ "Fidenza.readVectorFieldFromFile: malformatted file"
        toTuples (x:y:xys) = (x,y):toTuples xys
    let vectors = map (toTuples . (map read . words)) $ lines contents :: VectorField
    let (lenX,lenY) = (length $ head vectors, length vectors)
    return $ vectorFileNoiseFunc vectors (width,height) (lenX,lenY)

vectorFileNoiseFunc :: VectorField -> (Int,Int) -> (Int,Int) -> Vector -> Vector
vectorFileNoiseFunc vectors (width,height) (lenX,lenY) (x,y) = vec
  where (px,py) = (x/fromIntegral width, y/fromIntegral height)
        (px',py') = (px * (fromIntegral lenX-1), py * (fromIntegral lenY-1))
        (binX,binY) = (floor px', floor py')
        (vXY,vXY1,vX1Y,vX1Y1) = ( (!! binX) $ vectors !! binY 
                                , (!! binX) $ vectors !! (binY+1)
                                , (!! (binX+1)) $ vectors !! binY 
                                , (!! (binX+1)) $ vectors !! (binY+1)) 
        (u,v) = (px' - fromIntegral binX, py' - fromIntegral binY)
        lerpV (x1,y1) (x2,y2) w = (x1 + (x2 - x1) * w, y1 + (y2 - y1) * w)
        vec = lerpV (lerpV vXY vX1Y u)
                    (lerpV vXY1 vX1Y1 u) v


src/ParseArgs.hs
module ParseArgs ( Args(..)
                 , helpString
                 , parseArgs
                 ) where

import Codec.Picture (PixelRGBA8( .. ))

import qualified Data.Map as Map (Map, fromList, findWithDefault, map, (!), keys)

import VectorFieldGenerator (VectorFieldGenerator)
import Distribution (Distribution(..), readDistribution, nullDistribution)
import ColourSchemes (colourSchemes, readColour, readColourProbs)

data Args = Args
    { -- General options
      aSeed :: Int
    , aWidth :: Int
    , aHeight :: Int
    , aCollisionMargin :: Int
    , aPadding :: Int
    , aVectorFieldGenerator :: VectorFieldGenerator
    , aNumGenerationAttempts :: Int
    , aMaxSteps :: Int
    , aMaxCurves :: Int
    , aMinLength :: Int
    , aChunkSizes :: Distribution Int
    , aStopChunkingAt :: (Int,Int)
    , aSquareBlocks :: Float
    , aAvgBlockSize :: Float
      -- Curve generation options
    , aStepLength :: Float
    , _aCurveWidths :: [Float]
    , aWidths :: Distribution Float
    , aFertilities :: Distribution Int
    , aSkewAngles :: Distribution Float
      -- Drawing
    , aChunksOverlap :: Int
    , aBgColour :: PixelRGBA8
    , aColours :: Distribution PixelRGBA8
    , aRandomBoil :: Float
    , aRandomBoilSeed :: Int
    , aStrokeOrFill :: Int
    , aOutlineColour :: PixelRGBA8
    , aOutlineSize :: Float
    -- Softness
    , aDrawSoftly :: Bool
    , aSoftSeed :: Int
    , aSoftStepLength :: Int
    , aSoftMaxStrokes :: Int
    , aSoftNumStrokesWidthRatio :: Float
    , aSoftRandomOfsAlongPerp :: Float
    , aSoftRandomOfsAlongV :: Float
    , aSoftPerlinFreq :: Float
    , aSoftPerlinOfs :: Float
    } deriving stock (Show)

data ArgDesc = ArgDesc (String,String,String)
instance Show ArgDesc where
    show (ArgDesc (name,_default,desc)) =
        name ++ " - " ++ desc ++ ". Default " ++ _default

recognizedArgs :: Map.Map String ArgDesc
recognizedArgs = Map.fromList $ map (\(x,y,v) -> (x, ArgDesc (x,y,v))) [
                   ("seed","0","seed for the random generator")
                 , ("width","500","image width")
                 , ("height","600","image height")
                 , ("collisionMargin","2","")
                 , ("padding","20","")
                 , ("vectorFieldGenerator","PerlinNoise 0.001 -1000","")
                 , ("numGenerationAttempts","1000","")
                 , ("maxSteps","10000","")
                 , ("maxCurves","100","")
                 , ("minLength","15","")
                 , ("chunkSizes","[10,15]","")
                 , ("stopChunkingAt","(10000,10000)","")
                 , ("squareBlocks","0.0","")
                 , ("avgBlockSize","0.0","")
                   -- generation options
                 , ("stepLength","1","")
                 , ("widths","(2,60,4)","")
                 , ("fertilities","(25,1500,25)","")
                 , ("skewAngles","(-0.1,0.1,0.01)","")
                   -- drawing
                 , ("chunksOverlap","1","")
                 , ("colourScheme","0","")
                 , ("customBgColour","","")
                 , ("customColours","[]","")
                 , ("randomBoil","0","")
                 , ("randomBoilSeed","0","")
                 , ("strokeOrFill","1","")
                 , ("outlineColour","(10,10,10,255)","")
                 , ("outlineSize","1","")
                   -- softness
                 , ("drawSoftly","0","")
                 , ("softSeed","0","")
                 , ("softStepLength","11","")
                 , ("softMaxStrokes","220","")
                 , ("softNumStrokesWidthRatio","1.8","")
                 , ("softRandomOfsAlongPerp","2","")
                 , ("softRandomOfsAlongV","6","")
                 , ("softPerlinFreq","0.016","")
                 , ("softPerlinOfs","-300","-300")
                 ]
recognizedArgNames :: [String]
recognizedArgNames = map ("--"++) $ Map.keys recognizedArgs

helpString :: String
helpString = "Usage: (./Fidenza|cabal run Fidenza -- ) [ARGS] \n\
\ Ex:\n\
\   ./Fidenza --seed 1\n\
\   cabal run Fidenza -- --seed 1\n\
\ Args:\n" ++ (concat $ Map.map (("    "++) . (++"\n") . show) recognizedArgs)

parseArgs :: [String] -> Args
parseArgs args = parsedArgs
    where toTuples [] = []
          toTuples (x:[]) = error $ "Fidenza.parseArgs: unrecognized term " ++ x ++ ". \
                                    \Do you maybe have an unmatched flag or value?"
          toTuples (x:y:xys) = (x,y):toTuples xys
          argPairs = toTuples args
          unrecognizedArgs = filter
                               (not . (`elem` recognizedArgNames) . fst)
                               argPairs
          unrecognizedArgs' = drop 2 . concat $
                              map (\(k,v) -> ", " ++ k ++ " " ++ v) unrecognizedArgs
          argsMap = Map.fromList $ toTuples args
          getArg name = Map.findWithDefault _default ("--"++name) argsMap
              where (ArgDesc (_,_default,_)) = recognizedArgs Map.! name
          parsedArgs =
              case unrecognizedArgs of
                [] -> Args {..}
                    where aSeed = read $ getArg "seed"
                          aWidth = read $ getArg "width"
                          aHeight = read $ getArg "height"
                          aCollisionMargin = read $ getArg "collisionMargin"
                          aPadding = read $ getArg "padding"
                          aVectorFieldGenerator = read $ getArg "vectorFieldGenerator"
                          aNumGenerationAttempts = read $ getArg "numGenerationAttempts"
                          aMaxSteps = read $ getArg "maxSteps"
                          aMaxCurves = read $ getArg "maxCurves"
                          aMinLength = read $ getArg "minLength"
                          aChunkSizes = readDistribution $ getArg "chunkSizes"
                          aStopChunkingAt = read $ getArg "stopChunkingAt"
                          aSquareBlocks = read $ getArg "squareBlocks"
                          aAvgBlockSize = read $ getArg "avgBlockSize"
                          aStepLength = read $ getArg "stepLength"
                          _aCurveWidths = []
                          aWidths = readDistribution $ getArg "widths"
                          aFertilities = readDistribution $ getArg "fertilities"
                          aSkewAngles = readDistribution $ getArg "skewAngles"
                          aChunksOverlap = read $ getArg "chunksOverlap"
                          colourScheme = read $ getArg "colourScheme"
                          customColours = readColourProbs $ getArg "customColours"
                          aColours = if not $ nullDistribution customColours
                                     then customColours
                                     else snd $ colourSchemes !! colourScheme
                          aBgColour = if not $ null $ getArg "customBgColour"
                                      then readColour $ getArg "customBgColour"
                                      else fst $ colourSchemes !! colourScheme
                          aRandomBoil = read $ getArg "randomBoil"
                          aRandomBoilSeed = read $ getArg "randomBoilSeed"
                          aStrokeOrFill = read $ getArg "strokeOrFill"
                          aOutlineColour = readColour $ getArg "outlineColour"
                          aOutlineSize = read $ getArg "outlineSize"
                          aDrawSoftly = 1 == (read $ getArg "drawSoftly" :: Int)
                          aSoftSeed = read $ getArg "softSeed"
                          aSoftStepLength = read $ getArg "softStepLength"
                          aSoftMaxStrokes = read $ getArg "softMaxStrokes"
                          aSoftNumStrokesWidthRatio = read $ getArg "softNumStrokesWidthRatio"
                          aSoftRandomOfsAlongPerp = read $ getArg "softRandomOfsAlongPerp"
                          aSoftRandomOfsAlongV = read $ getArg "softRandomOfsAlongV"
                          aSoftPerlinFreq = read $ getArg "softPerlinFreq"
                          aSoftPerlinOfs = read $ getArg "softPerlinOfs"
                _  -> error $ "Fidenza.parseArgs: Unrecognized args: " ++ unrecognizedArgs'
